### PR TITLE
PRMT-4112 SonarCloud code smells

### PR DESCRIPTION
**Pull request by AndyFlintNHS**

### CODE REVIEW

File: src/main/java/uk/nhs/prm/repo/ehrtransferservice/config/SnsClientSpringConfiguration.java [hunk 1]
```diff
@@ -2,25 +2,32 @@
...
         return AmazonSNSClientBuilder.standard().withRegion(awsRegion).build();
     }
 
     @Bean
     public SnsClient snsClientV2() {
-        return SnsClient.builder().region(Region.of(awsRegion)).build();
+        return SnsClient.builder()
+                .region(Region.of(awsRegion))
+                .credentialsProvider(awsCredentialsProvider)
+                .build();
```
Comment (martin-nhs) [inline src/main/java/uk/nhs/prm/repo/ehrtransferservice/config/SnsClientSpringConfiguration.java:31]: My only concern, could this be problematic? https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/creating-clients.html#:~:text=Explicitly%20shut%20down%20clients%20when%20they%20are%20no%20longer%20needed%20to%20avoid%20resource%20leaks.

I don't believe this _will_ be an issue, just wanted to raise this. Maybe we can spike this
at some point in the future?

File: src/main/java/uk/nhs/prm/repo/ehrtransferservice/config/SnsClientSpringConfiguration.java [hunk 2]
```diff
@@ -2,25 +2,32 @@
...
+@RequiredArgsConstructor
 public class SnsClientSpringConfiguration {
     @Value("${aws.region}")
     private String awsRegion;
 
+    private AwsCredentialsProvider awsCredentialsProvider;
+
     @Bean
     public AmazonSNS snsClient() {
         return AmazonSNSClientBuilder.standard().withRegion(awsRegion).build();
```
Comment (martin-nhs) [inline src/main/java/uk/nhs/prm/repo/ehrtransferservice/config/SnsClientSpringConfiguration.java:23]: Could we break this into other lines if possible to keep it consistent?

```suggestion
        return AmazonSNSClientBuilder.standard()
        .withRegion(awsRegion)
        .build();
```

File: src/main/java/uk/nhs/prm/repo/ehrtransferservice/database/TransferStore.java [hunk 3]
```diff
@@ -33,8 +33,8 @@ public void createEhrTransfer(RepoIncomingEvent incomingEvent, String status) {
                             status,
                             getTimeNow(),
                             getTimeNow(),
-                            getLargeEhrCoreMessageId(),
-                            isActive()
+                            "", //This is a placeholder for the messageId from the LargeEhrCoreQueue,
```
Comment (martin-nhs) [inline src/main/java/uk/nhs/prm/repo/ehrtransferservice/database/TransferStore.java:36]: Would it be better to instead initialising this to an empty string here, create an additional constructor within `Transfer.java` that just initialises it there and does not require it?

File: src/main/java/uk/nhs/prm/repo/ehrtransferservice/gp2gp_message_models/MessageContent.java [hunk 4]
```diff
@@ -1,4 +1,4 @@
 package uk.nhs.prm.repo.ehrtransferservice.gp2gp_message_models;
 
-public abstract class MessageContent {
+public interface MessageContent {
```
Comment (martin-nhs) [inline src/main/java/uk/nhs/prm/repo/ehrtransferservice/gp2gp_message_models/MessageContent.java:3]: Marker interfaces are seemingly looked at with a negative light in modern Java, it's likely to point to a code smell! Maybe we can revisit this in the future and take a new architectural approach 😄 

Sources:

https://www.baeldung.com/java-marker-interfaces#:~:text=A%20marker%20interface%20is%20also,we%20should%20use%20them%20carefully.

File: src/main/java/uk/nhs/prm/repo/ehrtransferservice/handlers/NegativeAcknowledgementHandler.java [hunk 5]
```diff
@@ -21,14 +21,21 @@ public class NegativeAcknowledgementHandler {
...
+
+        transferStore.handleEhrTransferStateUpdate(
+                conversationId.toString(),
+                transferStore.findTransfer(conversationId.toString()).getNemsMessageId(),
+                createState(acknowledgement),
+                isActive);
+
+        publishTransferCompleteEvent(
+                transferStore.findTransfer(conversationId.toString()),
+                conversationId);
```
Comment (martin-nhs) [inline src/main/java/uk/nhs/prm/repo/ehrtransferservice/handlers/NegativeAcknowledgementHandler.java:38]: Another nit picky one - sorry!

```suggestion
        publishTransferCompleteEvent(
                transferStore.findTransfer(conversationId.toString()),
                conversationId
        );
```

File: src/main/java/uk/nhs/prm/repo/ehrtransferservice/handlers/NegativeAcknowledgementHandler.java [hunk 6]
```diff
@@ -21,14 +21,21 @@ public class NegativeAcknowledgementHandler {
...
         logFailureDetail(acknowledgement);
-        transferStore.handleEhrTransferStateUpdate(conversationId.toString(),
-                transferStore.findTransfer(conversationId.toString()).getNemsMessageId(), createState(acknowledgement), isActive);
-        publishTransferCompleteEvent(transferStore.findTransfer(conversationId.toString()), conversationId);
+
+        transferStore.handleEhrTransferStateUpdate(
+                conversationId.toString(),
+                transferStore.findTransfer(conversationId.toString()).getNemsMessageId(),
+                createState(acknowledgement),
+                isActive);
```
Comment (martin-nhs) [inline src/main/java/uk/nhs/prm/repo/ehrtransferservice/handlers/NegativeAcknowledgementHandler.java:34]: Another nit picky one - sorry!

```suggestion
        transferStore.handleEhrTransferStateUpdate(
                conversationId.toString(),
                transferStore.findTransfer(conversationId.toString()).getNemsMessageId(),
                createState(acknowledgement),
                isActive
        );
```

**Comment by martin-nhs** [comment]
A general comment (again very much optional) - there is a comment within `Tracer.java`: `// oops, coming in backdoor? - chat with team`. I know it's outside the scope of this ticket, though if we can remove this that would be fantastic! 😄

File: src/test/java/uk/nhs/prm/repo/ehrtransferservice/handlers/EhrCompleteHandlerTest.java [hunk 7]
```diff
@@ -51,27 +51,27 @@ void setUp() {
...
         assertThrows(Exception.class, () -> ehrCompleteHandler.handleMessage(ehrCompleteEvent));
         verify(transferStore, never()).handleEhrTransferStateUpdate(conversationId.toString(),"some-nems-message-id","ACTION:EHR_TRANSFER_TO_REPO_COMPLETE", false);
     }
 
     @Test
-    public void shouldPublishTransferCompleteMessageToTransferCompleteTopic() throws Exception {
+    void shouldPublishTransferCompleteMessageToTransferCompleteTopic() throws Exception {
         var transferComplete = new TransferCompleteEvent("nemsEventLastUpdated", "some-ods-code", "SUSPENSION", "some-nems-message-id", "some-nhs-number");
 //        when(transferTrackerDbEntry.getNhsNumber()).thenReturn("some-nhs-number");
 //        when(transferTrackerDbEntry.getSourceGP()).thenReturn("some-ods-code");
```
Comment (martin-nhs) [inline src/test/java/uk/nhs/prm/repo/ehrtransferservice/handlers/EhrCompleteHandlerTest.java:77]: Should we remove these comments?

**Comment by martin-nhs** [comment]
Great job @AndyFlintNHS looks great 😄

File: src/main/java/uk/nhs/prm/repo/ehrtransferservice/database/TransferStore.java (hunk 3 above)
Comment (AndyFlintNHS) [reply src/main/java/uk/nhs/prm/repo/ehrtransferservice/database/TransferStore.java:36]: This is something I'd like to address properly later down the line, there is a very large number of constructor calls for Transfer and the vast majority use 'true' for isActive. For now I'll leave it as-is, but I agree, I don't like this approach either.

image

File: src/main/java/uk/nhs/prm/repo/ehrtransferservice/gp2gp_message_models/MessageContent.java (hunk 4 above)
Comment (AndyFlintNHS) [reply src/main/java/uk/nhs/prm/repo/ehrtransferservice/gp2gp_message_models/MessageContent.java:3]: Funnily enough, SonarCloud flagged the marker Abstract Class as a code smell and asked that we change it to an interface! I'm not a fan of them either, I'm sure there is some logic we can move into an abstract class, but the only thing that implements this interface is one of those messy webs of publicly visible fields.

I'll leave it for now, fully agree with your point!

File: src/test/java/uk/nhs/prm/repo/ehrtransferservice/handlers/EhrCompleteHandlerTest.java (hunk 7 above)
Comment (AndyFlintNHS) [reply src/test/java/uk/nhs/prm/repo/ehrtransferservice/handlers/EhrCompleteHandlerTest.java:77]: Not a fan of that 'oops coming in backdoor' comment. It appears it only gets used on an EHR request timeout, I'll remove the comment for that.

Comment (AndyFlintNHS) [reply src/test/java/uk/nhs/prm/repo/ehrtransferservice/handlers/EhrCompleteHandlerTest.java:77]: Also removed your suggested comments, good spot!

**Review by martin-nhs — APPROVED** [review]

**Pull request merged**